### PR TITLE
Add panel to display charts and reports for CD

### PR DESCRIPTION
**Pull request by mkzia**
## Description
This pull request adds the chart and report panel in the center director's novice view.

## Motivation and Context
To make it easier for novice users to quickly fetch their charts and reports. 

## Tests performed
I have added preliminary tests for the new rest API that was added. It tests only two cases: fetching a chart only and fetching  a report only. I will later add a test for fetching both a chart and report. Note that the upgrade path test will fail because I have added a new column to the reports table. The migration file is in place to add the new column, but for it to run, we have to change the version of XDMoD to 8.5.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [X ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: html/gui/js/report_builder/ReportsOverview.js [hunk 1]
```diff
@@ -233,7 +243,7 @@ XDMoD.ReportsOverview = Ext.extend(Ext.Panel,  {
 
          store: this.reportStore,
          //id: 'reportPool_queueGrid' + Ext.id(),
-
+         id: 'reportPool_queueGrid',
```
Comment (jpwhite4) [inline html/gui/js/report_builder/ReportsOverview.js]: Is it safe to add an id here?

File: open_xdmod/modules/xdmod/integration_tests/lib/Controllers/MetricExplorerTest.php [hunk 2]
```diff
@@ -193,4 +193,4 @@ public function chartDataProvider()
             array($emptyChart)
         );
     }
-}
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/integration_tests/lib/Controllers/MetricExplorerTest.php]: Please don't change unrelated files in a pull request - it makes it difficult to review the changes.

File: html/gui/js/modules/summary/DeveloperPortlet.js [hunk 3]
```diff
@@ -59,3 +59,152 @@ XDMoD.Modules.SummaryPortlets.DeveloperPortlet = Ext.extend(Ext.ux.Portlet, {
  * can be dynamically instantiated
  */
 Ext.reg('DeveloperPortlet', XDMoD.Modules.SummaryPortlets.DeveloperPortlet);
+
```
Comment (jpwhite4) [inline html/gui/js/modules/summary/DeveloperPortlet.js]: This needs to be its own file

File: html/gui/js/modules/summary/DeveloperPortlet.js [hunk 4]
```diff
@@ -59,3 +59,152 @@ XDMoD.Modules.SummaryPortlets.DeveloperPortlet = Ext.extend(Ext.ux.Portlet, {
...
+
+/**
+ * 
+ * Add Charts and Reports Portlet
+ * 
+ */
+
+
+var chartReportRecord = Ext.data.Record.create([
```
Comment (jpwhite4) [inline html/gui/js/modules/summary/DeveloperPortlet.js]: these should not be at global scope

File: tests/artifacts/xdmod-test-artifacts/xdmod/summary_controller_provider/output/cd_chartsreports_report_only.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+            "report_format": "Pdf",
+            "report_id": "3-1551212717",
+            "report_name": "TAS Report 1",
+            "report_schedule": "Once",
+            "report_title": "",
+            "ts": 1551211883,
+            "config": 0
+        }
+    ]
+}
```
Comment (jpwhite4) [inline tests/artifacts/xdmod-test-artifacts/xdmod/summary_controller_provider/output/cd_chartsreports_report_only.json]: need to be posix compliant

File: classes/Rest/Controllers/SummaryControllerProvider.php [hunk 6]
```diff
@@ -58,47 +56,43 @@ public function getPortlets(Request $request, Application $app)
...
+                'column' => $column,
             );
         }
 
         $presetCharts = Roles::getConfig($mostPrivilegedAcl->getName(), 'summary_charts');
 
-        foreach ($presetCharts as $index => $presetChart)
-        {
-            list($chartLocation, $column) = $layout->getLocation('PC' . $index);
+        foreach ($presetCharts as $index => $presetChart) {
```
Comment (jpwhite4) [inline classes/Rest/Controllers/SummaryControllerProvider.php]: Prefer to not change unrelated code as part of a pull request - it makes it difficult for a reviewer to see what functionailty has been changed .

File: classes/Rest/Controllers/SummaryControllerProvider.php [hunk 7]
```diff
@@ -140,13 +134,61 @@ public function getPortlets(Request $request, Application $app)
             'success' => true,
             'total' => count($summaryPortlets),
             'portalConfig' => array('columns' => $layout->getColumnCount()),
-            'data' => array_values($summaryPortlets)
+            'data' => array_values($summaryPortlets),
```
Comment (jpwhite4) [inline classes/Rest/Controllers/SummaryControllerProvider.php]: No trailing commas as per coding standard.

File: classes/Rest/Controllers/SummaryControllerProvider.php [hunk 8]
```diff
@@ -140,13 +134,61 @@ public function getPortlets(Request $request, Application $app)
...
+                $tmp['report_id'] = $report['report_id'];
+                $tmp['report_name'] = $report['report_name'];
+                $tmp['report_schedule'] = $report['report_schedule'];
+                $tmp['report_title'] = $report['report_title'];
+                $tmp['ts'] = $report['last_modified'];
+                $tmp['config'] = $idx; // use config field to store index of reports
+                $data2[] = $tmp;
+                ++$idx;
+            }
+            $data = array_merge($data, $data2);
```
Comment (jpwhite4) [inline classes/Rest/Controllers/SummaryControllerProvider.php]: What is the reason for the `$data2` variable? Why didn't you just append directly to `$data` thus saving the temporary array construction and the array_merge call?

File: html/gui/js/modules/NoviceUser.js [hunk 9]
```diff
@@ -79,6 +79,7 @@ Ext.extend(XDMoD.Module.Summary, XDMoD.PortalModule, {
                     this.each(function (record) {
                         var config = record.get('config');
 
+
```
Comment (jpwhite4) [inline html/gui/js/modules/NoviceUser.js]: Please don't change this file.

File: tests/artifacts/xdmod-test-artifacts/xdmod/summary_controller_provider/output/cd_chartsreports_chart_only.json [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+            "ts": 1551214179.775,
```
Comment (jpwhite4) [inline tests/artifacts/xdmod-test-artifacts/xdmod/summary_controller_provider/output/cd_chartsreports_chart_only.json]: Why is the timestamp included here? How are you guaranteeing that it is this exact timestamp?

File: open_xdmod/modules/xdmod/integration_tests/lib/Controllers/SummaryControllerProviderTest.php [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+        // Attempt to create the report.
+        $reportId = $this->createReport($data);
+
+        // Ensure we were successful
+        $this->assertTrue(isset($reportId), "Did not receive a report_id back from create_report");
+
+        // Get reports
+        $response = $this->helper->get('rest/v1/summary/chartsreports');
+
+        $expected = JSON::loadFile(
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/integration_tests/lib/Controllers/SummaryControllerProviderTest.php]: You should add an assert that checks that `$reportId` is actually in the list of reports that comes back from the chartreports API call.

File: tests/artifacts/xdmod-test-artifacts/xdmod/summary_controller_provider/output/cd_chartsreports_report_only.json [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    "data": [
+        {
+            "type": "Report",
+            "name": "TAS Report 1",
+            "chart_count": "2",
+            "charts_per_page": "1",
+            "creation_method": "Manual",
+            "report_delivery": "E-mail",
+            "report_format": "Pdf",
+            "report_id": "3-1551212717",
```
Comment (jpwhite4) [inline tests/artifacts/xdmod-test-artifacts/xdmod/summary_controller_provider/output/cd_chartsreports_report_only.json]: How can you have a report_id in the expected data? Isn't it autogenerated int he code from a call to time() and will be different every time?

File: html/gui/js/modules/summary/DeveloperPortlet.js [hunk 13]
```diff
@@ -59,3 +59,152 @@ XDMoD.Modules.SummaryPortlets.DeveloperPortlet = Ext.extend(Ext.ux.Portlet, {
...
+                    var config = Ext.util.JSON.decode(r.data.config);
+                    XDMoD.Module.MetricExplorer.setConfig(config, config.summary_index, Boolean(config.preset));
+                } else if (r.data.type == 'Report') {
+                    selModel.clearSelections();
+                    var tabPanel = Ext.getCmp('main_tab_panel');
+                    tabPanel.setActiveTab('report_generator');
+                    var queueGrid = Ext.getCmp('reportPool_queueGrid');
+                    if (queueGrid.rowToSelect !== undefined) {
+                        var selModel = queueGrid.getSelectionModel();
+                        selModel.selectRow(r.data.config);
```
Comment (jpwhite4) [inline html/gui/js/modules/summary/DeveloperPortlet.js]: You cannot select reports based on the row number since the row number for a chart can change due to user input. The report_id is the unique identifier for a report and this should be used to select it.

Comment (jpwhite4) [reply html/gui/js/modules/summary/DeveloperPortlet.js]: Also please can you test whether it is ok to fire the rowdblclick event. What happens if a different report is being edited already? What happens it the row corresponding to the report is scrolled out of the window?

File: html/gui/js/modules/summary/DeveloperPortlet.js [hunk 14]
```diff
@@ -59,3 +59,152 @@ XDMoD.Modules.SummaryPortlets.DeveloperPortlet = Ext.extend(Ext.ux.Portlet, {
...
+    // { name: 'ts', type: 'date', dateFormat: 'timestamp' },
+    {
+        name: 'ts', convert: function (v, rec) {
+            return Ext.util.Format.date(new Date(rec["ts"] * 1000).toString(), 'Y-m-d h:i:s')
+        }
+    },
+    { name: 'recordid', type: 'int' }
+]); // chartReportRecord
+
+var chartReportStore = new Ext.data.Store({
```
Comment (jpwhite4) [inline html/gui/js/modules/summary/DeveloperPortlet.js]: You can use an Ext.data.JsonStore and it automatically creates the Json reader.

File: html/gui/js/modules/summary/DeveloperPortlet.js [hunk 15]
```diff
@@ -59,3 +59,152 @@ XDMoD.Modules.SummaryPortlets.DeveloperPortlet = Ext.extend(Ext.ux.Portlet, {
...
+    restful: true,
+    method: 'GET',
```
Comment (jpwhite4) [inline html/gui/js/modules/summary/DeveloperPortlet.js]: method: 'GET' is not needed

File: html/gui/js/modules/summary/DeveloperPortlet.js [hunk 16]
```diff
@@ -59,3 +59,152 @@ XDMoD.Modules.SummaryPortlets.DeveloperPortlet = Ext.extend(Ext.ux.Portlet, {
...
+    validationEvent: false,
+    validateOnBlur: false,
+    trigger1Class: 'x-form-clear-trigger',
+    trigger2Class: 'x-form-search-trigger',
+    hideTrigger1: true,
+    enableKeyEvents: true,
+    emptyText: 'Search',
+    store: chartReportStore,
+    onTrigger1Click: function () {
+        XDMoD.TrackEvent('Metric Explorer', 'Cleared chart search field');
```
Comment (jpwhite4) [inline html/gui/js/modules/summary/DeveloperPortlet.js]: The `XDMoD.TrackEvent` was used for the google analytics. We do not currently use this and it should not be in new code.

File: classes/Rest/Controllers/SummaryControllerProvider.php [hunk 17]
```diff
@@ -58,47 +56,43 @@ public function getPortlets(Request $request, Application $app)
...
+                $chartLocation = 'FW'.$portlet['name'];
                 $column = -1;
             } else {
-                list($chartLocation, $column) = $layout->getLocation('PP' . $portlet['name']);
+                list($chartLocation, $column) = $layout->getLocation('PP'.$portlet['name']);
             }
 
             $summaryPortlets[$chartLocation] = array(
-                'name' => 'PP' . $portlet['name'],
+                'name' => 'PP'.$portlet['name'],
```
Comment (plessbd) [inline classes/Rest/Controllers/SummaryControllerProvider.php]: we usually leave spaces around the concatenation as it makes it easier to read, also is there a reason for the whitespace changes or is this just something your editor is doing?

**Comment by mkzia** [comment]
New branch pushed incorporating all the changes requested.
